### PR TITLE
feat(testworkflows): TKC-1765 - parallel execution (K6/Playwright/Cypress)

### PR DESCRIPTION
**Pull request by rangoo94**
## Pull request description 

* Add support for `parallel` step
   * Run parallel steps (`parallel`)
   * Transfer content inside (`parallel.transfer`), and transfer it back to original workflow (`parallel.fetch`)
   * Isolate their artifacts
   * Expose their statuses in real-time
   * Expose their logs afterwards
   * Support sync lock (when the step has `paused: true` clause, the parallel executor will wait until all will reach this state, and resume them at once) and rendezvous similar way
      * not K6-specific - it will work with any kind of tool

## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [x] tested locally
- [x] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

## Fixes

- https://linear.app/kubeshop/issue/TKC-1765/[testworkflows]-k6-parallelism

### CODE REVIEW

File: cmd/tcl/testworkflow-toolkit/common/tarball.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+		}
+		if filepath.IsAbs(header.Name) || relativeCheckRe.MatchString(filepath.ToSlash(header.Name)) {
+			return fmt.Errorf("unsafe file path in the tarball: %s", header.Name)
+		}
+
+		filePath := filepath.Join(dirPath, header.Name)
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			err := os.Mkdir(filePath, 0755)
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/common/tarball.go:117]: It's validated and cleaned up few lines above

File: cmd/tcl/testworkflow-toolkit/common/tarball.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+			if err != nil {
+				return errors.Wrapf(err, "%s: create directory", filePath)
+			}
+		case tar.TypeReg:
+			err := os.MkdirAll(filepath.Dir(filePath), 0755)
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/common/tarball.go:122]: It's validated and cleaned up few lines above

File: cmd/tcl/testworkflow-toolkit/common/tarball.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+			if err != nil {
+				return errors.Wrapf(err, "%s: create directory tree", filePath)
+			}
+			outFile, err := os.OpenFile(filePath, os.O_CREATE|os.O_RDWR|os.O_TRUNC, os.FileMode(header.Mode))
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/common/tarball.go:126]: It's validated and cleaned up few lines above

File: cmd/tcl/testworkflow-toolkit/common/tarball.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+				return errors.Wrapf(err, "%s: create file", filePath)
+			}
+			_, err = io.Copy(outFile, tarReader)
+			if err != nil {
+				_ = outFile.Close()
+				return errors.Wrapf(err, "%s: write file", filePath)
+			}
+			_ = outFile.Close()
+		case tar.TypeSymlink:
+			err := os.MkdirAll(filepath.Dir(filePath), 0755)
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/common/tarball.go:137]: It's validated and cleaned up few lines above

File: cmd/tcl/testworkflow-toolkit/common/tarball.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+			if err != nil {
+				return errors.Wrapf(err, "%s: create directory tree", filePath)
+			}
+			err = os.Symlink(header.Linkname, filePath)
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/common/tarball.go:141]: It's validated and cleaned up few lines above

[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/common/tarball.go:141]: It's validated and cleaned up few lines above

File: cmd/tcl/testworkflow-toolkit/common/tarball.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+			break
+		}
+		if err != nil {
+			return errors.Wrap(err, "get next entry from tarball")
+		}
+		if filepath.IsAbs(header.Name) || relativeCheckRe.MatchString(filepath.ToSlash(header.Name)) {
+			return fmt.Errorf("unsafe file path in the tarball: %s", header.Name)
+		}
+
+		filePath := filepath.Join(dirPath, header.Name)
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/common/tarball.go:113]: It's validated and cleaned up few lines above

File: cmd/tcl/testworkflow-toolkit/common/tarball.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/common/tarball.go:141]: It's validated and cleaned up few lines above

**Review by exu — APPROVED** [review]

File: cmd/tcl/testworkflow-toolkit/commands/parallel.go [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+			infos := make([]string, 0)
+			if params.MatrixCount > 1 {
+				infos = append(infos, fmt.Sprintf("%d combinations", params.MatrixCount))
+			}
+			if params.ShardCount > 1 {
+				infos = append(infos, fmt.Sprintf("sharded %d times", params.ShardCount))
+			}
+			parallelism := int64(parallel.Parallelism)
+			if parallelism <= 0 {
+				parallelism = 1000
```
Comment (vsukhin) [inline cmd/tcl/testworkflow-toolkit/commands/parallel.go:140]: looks like const

File: cmd/tcl/testworkflow-toolkit/commands/parallel.go [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+				if _, err = transferSrv.Listen(); err != nil {
+					ui.Fail(errors.Wrap(err, "failed to start transfer server"))
+				}
+				fmt.Printf("Transfer server started.\n")
+			}
+
+			// Validate if there is anything to run
+			if len(specs) == 0 {
+				fmt.Printf("nothing to run\n")
+				os.Exit(0)
```
Comment (vsukhin) [inline cmd/tcl/testworkflow-toolkit/commands/parallel.go:295]: somewhere ui package is used, there os and fmt

File: cmd/tcl/testworkflow-toolkit/commands/parallel.go [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+				if err != nil {
+					fmt.Printf("%d: failed to prepare resources: %s\n", index, err.Error())
+					return false
+				}
+				defer func() {
+					// Save logs
+					reader, writer := io.Pipe()
+					filePath := fmt.Sprintf("logs/%d.log", index)
+					ctrl, err := testworkflowcontroller.New(context.Background(), clientSet, env.Namespace(), id, scheduledAt, testworkflowcontroller.ControllerOptions{
+						Timeout: 120 * time.Second,
```
Comment (vsukhin) [inline cmd/tcl/testworkflow-toolkit/commands/parallel.go:350]: also internal const

File: cmd/tcl/testworkflow-init/data/global.go [hunk 10]
```diff
@@ -8,11 +8,16 @@
...
+import (
+	"os"
 
-var baseTestWorkflowMachine = expressionstcl.CombinedMachines(EnvMachine, StateMachine, FileMachine)
+	"github.com/kubeshop/testkube/pkg/tcl/expressionstcl"
+	"github.com/kubeshop/testkube/pkg/tcl/expressionstcl/libs"
+)
 
 func GetBaseTestWorkflowMachine() expressionstcl.Machine {
+	var wd, _ = os.Getwd()
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-init/data/global.go:19]: I think that it's better to ignore it, as it's a rather uncommon behavior, but I'm not 100% sure

**Review by vsukhin — APPROVED** [review]

File: cmd/tcl/testworkflow-toolkit/commands/parallel.go [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+
+			if success.Load() == params.Count {
+				fmt.Printf("Successfully finished %d workers.\n", params.Count)
+			} else {
+				fmt.Printf("Failed to finish %d out of %d expected workers.\n", params.Count-success.Load(), params.Count)
+				os.Exit(1)
+			}
+		},
+	}
+
```
[the remark this answers is not in the document]
Comment (rangoo94) [reply cmd/tcl/testworkflow-toolkit/commands/parallel.go:339]: I've just split plenty of the code outside, so should be much better 👍

**Pull request merged**